clang/lib/Driver/Driver.cpp

```diff
@@ -3858,13 +3871,16 @@ class OffloadingActionBuilder final {
         Action *A(HostAction);
         // Only check for FPGA device information when using fpga SubArch.
         if (Args.hasArg(options::OPT_fintelfpga) &&
-            HostAction->getType() != types::TY_FPGA_AOCR &&
-            HostAction->getType() != types::TY_FPGA_AOCX &&
             !(HostAction->getType() == types::TY_Object &&
               isObjectFile(InputName))) {
-          if (HasFPGADeviceBinary(C, InputArg->getAsString(Args), true))
+          // Type FPGA aoco is a special case for -foffload-static-lib.
+          if (HostAction->getType() == types::TY_FPGA_AOCO) {
+            if (!hasFPGABinary(C, InputName, types::TY_FPGA_AOCO))
+              return false;
+            A = C.MakeAction<InputAction>(*InputArg, types::TY_FPGA_AOCO);
+          } else if (hasFPGABinary(C, InputName, types::TY_FPGA_AOCX))
             A = C.MakeAction<InputAction>(*InputArg, types::TY_FPGA_AOCX);
-          else if (HasFPGADeviceBinary(C, InputArg->getAsString(Args)))
+          else if (hasFPGABinary(C, InputName, types::TY_FPGA_AOCR))
```

Comment:
Maybe a switch statement on the pre-extracted `HostAction->getType()` would work?

Reply:
I don't think a switch statement will work here.  We are just doing checks for a particular section in the input file and adding an action based on if it is found or not.  Perhaps the hasFPGABinary function can be improved to check all known types and return the type if found?

Reply:
Oh, I've simply misread the code - thanks for the explanation! +1 on leaving as is due to better flexibility